File: collect_app/src/main/java/org/odk/collect/android/formmanagement/FormDownloader.java
package org.odk.collect.android.formmanagement;

import org.odk.collect.forms.FormSourceException;

import java.util.function.Supplier;

import javax.annotation.Nullable;

public interface FormDownloader {

    void downloadForm(ServerFormDetails form, @Nullable ProgressReporter progressReporter, @Nullable Supplier<Boolean> isCancelled) throws FormSourceException;

    interface ProgressReporter {
        void onDownloadingMediaFile(int count);
    }
}

File: collect_app/src/main/java/org/odk/collect/android/formmanagement/FormDownloader.kt
package org.odk.collect.android.formmanagement

import java.util.function.Supplier

interface FormDownloader {

    @Throws(FormDownloadException::class)
    fun downloadForm(
        form: ServerFormDetails?,
        progressReporter: ProgressReporter?,
        isCancelled: Supplier<Boolean?>?
    )

    interface ProgressReporter {
        fun onDownloadingMediaFile(count: Int)
    }
}
